### ISSUE TITLE
LLVM Memory Model needs more rigor to avoid undesired optimization results

### ISSUE THREAD

**Issue by RalfJung** [llvm/llvm-project#34577]
|  |  |
| --- | --- |
| Bugzilla Link | [35229](https://llvm.org/bz35229) |
| Version | 5.0 |
| OS | Linux |
| Depends On | llvm/llvm-project#33896  |
| Blocks | llvm/llvm-project#39717  |
| CC | @comex,@efriedma-quic,@hfinkel,@jrmuizel,@jplatte,@aqjune,@sunfishcode,@nunoplopes,@regehr,@rnk |

## Extended Description 
Clang/LLVM currently miscompiles the following program:
```cpp
// gvnbug.c
#include <stdio.h>
#include <stdint.h>

int foo();

void test(int* gp1, int* gp2)
{
  int g = 0;
  int a = 0, b = 0;
  int x = 7777, y = 6666; // also try swapping these
  int* p = &g;
  int* q = &g;
  int* r = &g;

  if (foo()) {
    a = 1;
    p = &y+1;
    q = &x;
  }

  *gp1 = (uintptr_t)p+1;
  if (q == p) {
    b = 1;
    *gp2 = (uintptr_t)q+1;
    r = q;
  }
  *r = 42;

  printf("a = %d, b = %d, x = %d\n", a, b, x);
}

int main() {
  int gp1 = 0;
  int gp2 = 0;

  test(&gp1, &gp2);

  return 0;
}

// aux.c
int foo() { return 1; }
```
```console
$ clang-5.0 aux.c gvnbug.c -o gvnbug -O3 && ./gvnbug 
a = 1, b = 1, x = 7777
```
This result is not allowed.  If `a` and `b` are both `1`, the branch `q == p` must have been taken, so `r` was set to `&x` (via `q`), so `x` cannot be `7777`.

I think this issue has already come up in https://github.com/llvm/llvm-project/issues/33896, but so far there was no example showing that the bug arises independent of the incorrect inttoptr-simplification.

What is happening here (if my analysis is correct) is that GVN sees the equality `q == p` and uses that to replace `q` by `p` in the then-branch.  Next, LLVM notices that because `p` is derived from `y`, writing to `r` (which will either have value `&g` or `p` in the line where the assignment happens) cannot possibly affect `x`, and hence the initial value of `x` can be propagated into the `printf`.  GVN is wrong to perform this kind of replacement; just because the bit representations of two pointers are equal, that doesn't mean that their provenance information is equal.

Test case by Gil Hur.

**Comment by aqjune** [comment]
Hi Kosta,

If I understand correctly, it is more like a bug of PRE incorrectly assuming that r.pointer1 is equivalent to et.pointer1 at the first iteration of the loop.

After -O1, the bitcode looks like this:

```llvm
pointer = calloc(1, 4)
et = entry(pointer, 0)
%1 = et.pointer1

if (%1 != null) { // Note that %1 is always null, so this is never taken.
  %2 = pointer[0] // always 0
  do {
    p = phi(0, %select)
    inc = p + 1
    select = (inc == %2) ? 0 : inc // always inc
    et2 = entry(pointer, select)
  } while(et2.pointer1 != null)
}

i2 = phi(0, select)
pointer[i2] = 1
r = entry(pointer, i2)
printf(.., r.pointer1, r.pointer2)
```

GVN-PRE replaces r.pointer1 with phi(et.pointer1, et2.pointer1), which is incorrect because `pointer[i2] = 1` makes `r` and `et` different.

The incorrect replacement may have been fired due to the existence of the null comparison, but I guess it is not related with our bug (which replaces a pointer with another one with different origin).

**Comment by RalfJung** [comment]
> TL;DR Given the current state of the world and LLVM memory model, it's unclear what is okay and not.

For what it's worth, while the LLVM memory model is indeed very unclear, I have not even seen a rough proposal of a sketch of a model that would allow this optimization. LLVM has a definition of "based on" for pointers that is quite clearly relevant, and the optimization in the OP can replace a pointer "based on" X by a pointer "based on" Y -- which is clearly not equivalent, so this is clearly not something an optimization is allowed to do. Many other questions remain open, but I think this one question we can answer with certainty: replacing one pointer value by another just because their addresses are equal is incorrect in a language like LLVM IR where "based on" for pointers is a deeply relevant notion (and not just an analysis artifact).

It'd be great if LLVM could at least offer the option for frontends to opt-out of this clearly incorrect optimization (similar to the `--disable-i2p-p2i-opt` flag that was added recently -- arguably this pointer replacement is worse since it will also affect code that doesn't even do any funny pointer-integer things).

**Comment by aqjune** [comment]
We have `llvm::canReplacePointersIfEqual` (https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/Analysis/Loads.h#L177, https://github.com/llvm/llvm-project/blob/main/llvm/lib/Analysis/Loads.cpp#L646), and I believe GVN/NewGVN/CVP would be the right place to use this helper.

**Comment by RalfJung** [comment]
Yeah if we could just make that always return `false` for Rust that'd be a good start I think. ;)

**Comment by RalfJung** [comment]
Here is [another example of the same bug](https://godbolt.org/z/qYPosGvh9) due to [Matthew House](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=105294#c2). This actually gets miscompiled by both clang and GCC.

**Comment by RalfJung** [comment]
Turns out there has been some progress here about a year ago: https://github.com/llvm/llvm-project/pull/82458 disables the incorrect optimization for many cases. There are some cave-outs unfortunately; not sure if there is any chance of those being removed eventually?

However, there is one comment in the PR which doesn't seem quite correct to me. @nunoplopes wrote
> Where use(p) is any of icmp or ptrtoint.
This is correct because these users only care about the pointer address not its provenance.

`ptrtoint` *does* care about the provenance of the pointer -- it has to mark that provenance as exposed so that future `inttoptr` can pick it up. Attempts to treat `ptrtoint` as a pure operation that just returns the address and ignores the provenance are [bound to fail](https://www.ralfj.de/blog/2022/04/11/provenance-exposed.html).

**Comment by RalfJung** [comment]
https://github.com/llvm/llvm-project/pull/151953 says GVN already handles this correctly when propagating equalities so... has this been fixed? @nikic

**Comment by nikic** [comment]
It's probably sufficiently fixed to close this issue. The semantics here are now clear and specified in LangRef (icmp ptr only compares pointer address bits) and their consequences are well-understood and implemented, with a caveat.

In particular, we still have some pragmatic carve-outs to limit the impact of this fix, and one of them is that replacements with dereferenceable globals are allowed. That case is problematic particularly in conjunction with noalias pointers. However, there is a separate issue to track that specific problem: https://github.com/llvm/llvm-project/issues/59679

**Comment by RalfJung** [comment]
> The semantics here are not clear and specified in LangRef 

Did you mean "now clear"?

**Comment by RalfJung** [comment]
https://github.com/llvm/llvm-project/issues/220206 tracks more leftover carve-outs that are still fundamentally the same unsoundness as this issue.
